src/sparseml/optim/manager.py

```diff
@@ -207,9 +207,28 @@ def compose_staged(
         if not keep_original_epochs:
             # update additional modifier epochs
             base_end_epoch = base_recipe.max_epochs
+
+            # make sure that for the modifiers in base_stages
+            # with the initial attribute `end_epoch` = -1,
+            # this attribute value is replaced with `base_end_epoch`
+            for base_modifiers in base_stages.values():
+                for base_modifier in base_modifiers:
+                    if (
+                        hasattr(base_modifier, "end_epoch")
+                        and base_modifier.end_epoch == -1
+                    ):
+                        base_modifier._init_end = base_end_epoch
```

Comment:
where is _init_end used?

Reply:
Without _init_end modification, we get ValueError at L707 `optim/modifier.py`:
```
        if self._end_comparator is None and self._end_epoch != self._init_end:
            raise ValueError(
                "end_epoch of {} must be equal the init value of {} for {}".format(
                    self._end_epoch, self._init_end, self.__class__.__name__
                )
            )
```

Reply:
looks good, one thing we need to eventually consider is some modifiers we may want to have automatically end at the end of a stage if indefinite (maybe constant pruning), but some should stick around forever such as quantization. I think we're ok for now, but as this gets more used and we hit corner cases and new modifiers this may come up

Reply:
Hmm, this is a super interesting corner case. I like the idea of handling them progressively. At the same time, we may now start documenting them and logging them into the backlog, what do you think?

Reply:
@bfineran what's the context for the conditional on optim/modifier.py::707? Because it seems it's designed to prevent modifications of the end epoch and I'm wondering if it's an anti pattern to work around the conditional rather than to modify/delete it

Reply:
@KSGulin very good point

Reply:
idea here was if something doesn't have an end comparator then the end epoch likely should be -1 or something else that shouldn't be set which is why this check is in place.